Port clientside encryption to mono-dll

jaschrep-msft opened the pull request:
Ported client-side encryption work from long-diverged branch. Was an inheritance approach, is now part of the main package internals.
Ported over internal WindowStream from stg73base, which allows us to prematurely end streams.
Removed internal RollingBufferStream, as uploads no longer require seekable streams as input. Upload paths that handle non-seekable streams already have their own buffer strategy.

jaschrep-msft committed:
Port clientside encryption to mono-dll

Ported blob client-side encryption work from long-diverged branch. Was an
inheritance approach, is now part of the main package internals.
Ported over WindowStream from stg73base, which allows us to prematurely
end streams.
Removed RollingBufferStream, as uploads no longer require seekable
streams as input.

jaschrep-msft commented:
Ported over blobs. Coming up in this PR, porting queues, extra tests for blob and queue, introduction of ThrowForMissingDecryptionKey flag, and bug fix on queue plaintext message deserialization.

jaschrep-msft committed:
Removed crypto package from ci.yml

jaschrep-msft committed:
regenerate/export-api

jaschrep-msft committed:
Ported clientside encryption for queues

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptedBlobRange.cs:
```diff
@@ -0,0 +1,72 @@
+        /// </summary>
+        public HttpRange AdjustedRange { get; }
+
+        public EncryptedBlobRange(HttpRange originalRange)
```
I'm not sure if constructor is right place for such heavy operation.
Don't we just need some kind of `HttpRange ExtendRange(HttpRange source)` transformation ?

jaschrep-msft replied:
We can. This was ported over from I think Java track 2, so we can change it to suit our needs.

jaschrep-msft replied:
Addressed in upcoming push

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/ClientSideEncryptionAlgorithm.cs:
```diff
@@ -0,0 +1,68 @@
+    /// <summary>
+    /// Specifies the encryption algorithm used to encrypt and decrypt a blob.
+    /// </summary>
+    internal readonly struct ClientSideEncryptionAlgorithm
```
should this be enum ?

jaschrep-msft replied:
We can. It would be just as much code maintenance, though, because I'd have to write the custom serialize/deserialize logic. the enum name `AES_CBC_256` won't get through OneSDK code review (I believe they moved me to this).

kasobol-msft replied:
If this is internal only we may consider just using string constants then.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/ClientsideEncryptionOptions.cs:
```diff
@@ -0,0 +1,63 @@
+    /// <summary>
+    /// Provides the client configuration options for connecting to Azure Blob using clientside encryption.
+    /// </summary>
+    public class ClientSideEncryptionOptions
```
Are we adding flag that controls whether exception is thrown or not if client can't decrypt separately?

jaschrep-msft replied:
Yes. Not pushed yet, but I'm almost done testing.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/MockKeyEncryptionKey.cs:
```diff
@@ -0,0 +1,164 @@
+    /// <summary>
+    /// Mock for a key encryption key. Not meant for production use.
+    /// </summary>
+    internal class MockKeyEncryptionKey : IKeyEncryptionKey, IKeyEncryptionKeyResolver,
+        IKey, IKeyResolver // for track 1 compatibility tests
```
I'd rather break it down into 4 separate types. Unless we have a good reason to create [God object](https://en.wikipedia.org/wiki/God_object)

jaschrep-msft replied:
It makes cleaner tests. This improved my test-writing ability substantially.

I wouldn't consider this a god class, as it's really only doing two small actions and exposing them repeatedly through a large variety of interfaces understood by track 1 and track 2. This ensures we are doing the same operations across track 1 and 2 in our testing so we can know that we're emulating comparable situations.

kasobol-msft replied:
Then the class name is wrong (and honestly I'm not sure how to name it, which is another hint this should be spitted). I'm not sure how test-writing ability is improved by combining loosely related functionalities - if it's creating 4 fields/vars in test vs 1 then I think it's not a big improvement (ink is cheap on the computer).
I've seen test aids like that in the past (that were trying to abstract too much). It usually ended up not well in long run. (coupling functionality is usually more expensive to change in long ran than not having to write extra variable names).
Not a very big deal though at this point, we can break it later when we feel more pain. Make your call here.

jaschrep-msft replied:
We discussed elsewhere to make better use of Moq anyway. I'll go that route, having functions that do the work being placed onto different mocks. Will accomplish my goal of maintaining code to do these mock operations in one place.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/ClientSideEncryptionTests.cs:
```diff
@@ -0,0 +1,388 @@
+
+        [Test]
+        [LiveOnly] // cannot seed content encryption key
+        [Ignore("stress test")]
```
Does this just fetch and decrypt 10 times 10MB? If so maybe we could enable this.

jaschrep-msft replied:
This test alone runs quite a while last I ran it (that was a dinky laptop, though). People pushed me for an ignore attribute. I can check again if it's viable to run, especially since it's live only.

kasobol-msft replied:
If you manage to run it reasonably fast then consider changing name of this test.

jaschrep-msft replied:
This test is about 16x longer than the rest of the clientside encryption tests combined, passing 8 minutes.

seanmcc-msft replied:
8 min isn't going to fly as a test we run locally, I vote we continue to leave this test marked as ignore.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/ClientSideEncryptionTests.cs:
```diff
@@ -0,0 +1,388 @@
+        [Ignore("stress test")]
+        public async Task StressAsync()
+        {
+            static async Task<byte[]> RoundTripDataHelper(BlobClient client, byte[] data)
```
nit: maybe RoundTirpDataVerifier?

Suffixes like util, helper don't bring much extra understanding.
https://softwareengineering.stackexchange.com/questions/247267/what-is-a-helper-is-it-a-design-pattern-is-it-an-algorithm

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/ClientSideEncryptionTests.cs:
```diff
@@ -0,0 +1,388 @@
+
+        #region Utility
+
+        private byte[] LocalManualEncryption(byte[] data, byte[] key, byte[] iv)
```
nit: method naming should be verb-ish. `EncryptLocally` or `DoLocalEncryption`. noun-ish terms are reserved for types. btw. what does manual mean?

jaschrep-msft replied:
It means I'm invoking the encryption myself, instead of going through the client's automated processes. Maybe explicit is a better word than manual?

gapra-msft replied:
Maybe bypass may be a candidate term to consider? Is that what you're trying to do?

kasobol-msft replied:
`Independent`/`External`/`Baseline`? 

Manual indicates human involvement in the process.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/ClientSideEncryptionTests.cs:
```diff
@@ -0,0 +1,388 @@
+        }
+
+
+        #region Utility
```
nit: I'm not sure what this region means. besides it's not closed.

jaschrep-msft replied:
It is closed, write above the first test.

Region is intended to separate tests from methods that help run the tests.

kasobol-msft replied:
okay. I think vertical placement (tests near tests, aids near aids) is good enough. I'd avoid `Utility` naming if you want to keep region

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/ContentRange.cs:
```diff
@@ -0,0 +1,149 @@
+            }
+
+            /// <summary>
+            /// AES-CBC using a 256 bit key.
```
this class doesn't seem to be dependent on algorithm. this information may find better home.

jaschrep-msft replied:
Copy/paste error. Property is necessary but comment is irrelevant.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:
```diff
@@ -1037,20 +1068,54 @@ public BlobClient(Uri blobUri, TokenCredential credential, BlobClientOptions opt
+        /// </param>
         /// <returns>Transformed content stream and metadata.</returns>
-        internal virtual (Stream, Metadata) TransformContent(Stream content, Metadata metadata)
+        private async Task<(Stream, Metadata)> ClientSideEncryptInternal(
```
Maybe we could keep the `TransformContent` idea and still inject encryption just make it all internal? that way we'd keep crypto related stuff in other class/namespace without cluttering blobclient.

jaschrep-msft replied:
I've pulled out all the crypto transformations, but have left in the explicit call to an encryption method instead of a generic TransformContent. Upcoming push.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -686,6 +709,15 @@ private void SetNameFieldsIfNull()
+                    // we already return a nonseekable stream; returning a crypto stream is fine
+                    if (UsingClientSideEncryption)
+                    {
+                        stream = await ClientSideDecryptInternal(stream, response.Value.Metadata, encryptedRange.OriginalRange, response.Value.ContentRange, async, cancellationToken).ConfigureAwait(false);
```
maybe we could make this extension and move out crypto code from this client ?

jaschrep-msft replied:
Created new classes to handle the translation between clients and service-agnostic encryption/decryption logic. Upcoming push.

kasobol-msft reviewed (commented):
The publicly facing part looks good (I left ~2 questions for that). 
I think we should work little bit on cleaning up the implementation.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -2950,6 +2982,117 @@ private void SetNameFieldsIfNull()
+                return content; // TODO readjust range
+            }
+
+            bool ivInStream = originalRange.Offset >= 16;
```
Initialization vector. IV is the common abbreviation, even in the .NET API.

kasobol-msft replied:
okay. looked little bit like a typo.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs:
```diff
@@ -252,6 +264,7 @@ internal BlobServiceClient(Uri serviceUri, HttpPipelinePolicy authentication, Bl
                   options?.Version ?? BlobClientOptions.LatestVersion,
                   new ClientDiagnostics(options),
                   options?.CustomerProvidedKey,
+                  options?._clientSideEncryptionOptions?.Clone(),
```
I would love all of our options classes to be immutable. I feel it's good to present in the same way as our client options classes, which are get/set properties.

Really, I'm handling this the same way we handle properties in our BlobClientOptions. I'm only keeping them grouped in an object instead of pulling each one out.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/WrappedKey.cs:
```diff
@@ -0,0 +1,26 @@
+    /// <summary>
+    /// Represents the envelope key details stored on the service.
+    /// </summary>
+    internal class WrappedKey
```
This is the json schema of clientside encryption.

kasobol-msft replied:
By looking at other comment I'd name it `KeyEnvelope` . Class name isn't that relevant for json serialization/deserialization - I'd rather better understand code when I'm reading it.

kasobol-msft replied:
Btw. You may elaborate little bit in comment in this class to reflect what you've explained in comment below - this really helps understanding this.

jaschrep-msft replied:
Let me know if the comments are clarified.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/WrappedKey.cs:
```diff
@@ -0,0 +1,26 @@
+        public string KeyId { get; set; }
+
+        /// <summary>
+        /// The encrypted content encryption key.
```
This is the wrapped key, per an envelope strategy. So this is the real key that encrypted the data, but the customer has used a keywrap algorithm on it with their personally managed key.

kasobol-msft replied:
Sounds good. Please make comments explain what you just said.

jaschrep-msft replied:
Let me know if comments are clarified.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,102 @@
+        /// <summary>
+        /// Singleton string identifying this encryption library.
+        /// </summary>
+        private static string AgentString { get; } = new Func<string>(() =>
```
Will inline property assignment allow met to call a function defined elsewhere? I thought there was a compiler restriction for some reason. I can double check that.

kasobol-msft replied:
I usually seen just named static function created for that purpose.

static string ComputeX()

static string X { get; } = ComputeX();

jaschrep-msft replied:
Addressed in next push.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,102 @@
+                    EncryptionAlgorithm = ClientSideEncryptionAlgorithm.AesCbc256,
+                    Protocol = ClientSideEncryptionVersion.V1_0
+                },
+                KeyWrappingMetadata = new Dictionary<string, string>()
+                {
+                    { EncryptionConstants.AgentMetadataKey, AgentString }
+                },
+                WrappedContentKey = new WrappedKey()
+                {
+                    Algorithm = keyWrapAlgorithm,
+                    EncryptedKey = async
+                        ? await keyEncryptionKey.WrapKeyAsync(keyWrapAlgorithm, contentEncryptionKey, cancellationToken).ConfigureAwait(false)
+                        : keyEncryptionKey.WrapKey(keyWrapAlgorithm, contentEncryptionKey, cancellationToken),
+                    KeyId = keyEncryptionKey.KeyId
```
Wrapped in this context means that the key itself has been encrypted using a keywrap algorithm. So it does actually convey something.

The schema does need to be track one compatible. This is serialized and stored on the metadata, which track 1 needs to be able to deserialize and read.

kasobol-msft replied:
That'd be first time `wrap` means something other than generic `wrapper` pattern. Thanks for explanation.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,102 @@
+        /// </summary>
+        /// <param name="json">JSON to deserialize.</param>
+        /// <returns></returns>
+        public static EncryptionData Deserialize(string json)
+            => EncryptionDataSerializer.Deserialize(json);
```
To match the `Serialize()` instance method. My understanding is that having an instance serialize method and static deserialize method is a good pattern to have on an object intended to be serialized. Should I not be doing this?

kasobol-msft replied:
I usually avoid adding serialize and deserialize method into POCO. The reason is that there might be many serialization strategies at some point (be open for extension). 
So I'd leave the `EncryptionData` as pure model class and have serializer/deserializer as external service/extension.

jaschrep-msft replied:
Github needs a reaction or a flag to say "addressed in upcoming push," because I don't want to mark it as resolved when I haven't pushed it or in case it's not actually what the commenter wanted, but continually typing this is annoying.

BTW, addressed in upcoming push.

gapra-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -914,6 +914,11 @@ public partial class UserDelegationKey
 }
 namespace Azure.Storage.Blobs.Specialized
 {
+    public partial class AdvancedBlobClientOptions : Azure.Storage.Blobs.BlobClientOptions
```
I like Extended

jaschrep-msft replied:
I can go either way. Can we get @tg-msft's input on naming decisions?

jaschrep-msft replied:
I moved to Extended. We'll let final API review catch the name if it's unwanted.

gapra-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -2950,6 +2982,117 @@ private void SetNameFieldsIfNull()
+            EncryptionData encryptionData = GetAndValidateEncryptionDataOrDefault(metadata);
+            if (encryptionData == default)
+            {
+                return content; // TODO readjust range
```
Is this still relevant?

jaschrep-msft replied:
Yes. Yes it was. Addressed in upcoming push.

gapra-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -188,7 +188,14 @@ public PageBlobClient(Uri blobUri, TokenCredential credential, BlobClientOptions
             CustomerProvidedKey? customerProvidedKey,
             string encryptionScope)
-            : base(blobUri, pipeline, version, clientDiagnostics, customerProvidedKey, encryptionScope)
+            : base(
```
Why do we throw in the toBlockBlobClient method but not in PageBlob/AppendBlob? Did I miss something?

jaschrep-msft replied:
Added to the extension methods at the bottom of the files, as in blockblobclient. Probably have to address this with public constructors as well.

gapra-msft replied:
Should the TypeNotSupported method pass in typeof(PageBlobClient)? Otherwise looks good.

gapra-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionDataSerializer.cs:
```diff
@@ -0,0 +1,165 @@
+using System.Text;
+using System.Text.Json;
+
+namespace Azure.Storage.Cryptography.Models
```
It'd be nice to have some comments on some of these methods to visualize what's happening.

jaschrep-msft replied:
Added in upcoming push.

gapra-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Utility.cs:
```diff
@@ -0,0 +1,295 @@
+
+namespace Azure.Storage.Cryptography
+{
+    internal static class Utility
```
Maybe you can split this into some sort of Encrypt and Decrypt classes that contain those specific methods?

jaschrep-msft replied:
Upcoming commit: split into ClientSideEncryptor, ClientSideDecryptor, and ClientSideEncryptionOptionsExtensions.

gapra-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/Errors.Clients.cs:
```diff
@@ -84,5 +85,16 @@ public static void VerifyHttpsTokenAuth(Uri uri)
+
+        public static class ClientSideEncryption
+        {
+            public static InvalidOperationException TypeNotSupported(Type type)
```
What is the difference between this and EncryptionErrors? Is one more related to client side encryption itself whereas this is more client related?

jaschrep-msft replied:
Leftover from split packages. Will cleanup before merge.

gapra-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClient.cs:
```diff
@@ -2043,5 +2092,59 @@ private void SetNameFieldsIfNull()
         }
         #endregion UpdateMessage
+
+        private async Task<string> ClientSideEncryptInternal(string messageToUpload, bool async, CancellationToken cancellationToken)
```
This may not be too important since the queue stuff is rarely touched, but you could consider moving these methods to a separate class to avoid cluttering the queue client.

jaschrep-msft replied:
addressed in upcoming push

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -914,6 +914,11 @@ public partial class UserDelegationKey
 {
+    public partial class AdvancedBlobClientOptions : Azure.Storage.Blobs.BlobClientOptions
+    {
+        public AdvancedBlobClientOptions(Azure.Storage.Blobs.BlobClientOptions.ServiceVersion version = Azure.Storage.Blobs.BlobClientOptions.ServiceVersion.V2019_07_07) : base (default(Azure.Storage.Blobs.BlobClientOptions.ServiceVersion)) { }
```
The actual code uses LatestVersion. Generated API file converts that into this. If you look at the base type in this file it looks exactly the same.
```csharp
public BlobClientOptions(Azure.Storage.Blobs.BlobClientOptions.ServiceVersion version = Azure.Storage.Blobs.BlobClientOptions.ServiceVersion.V2019_07_07) { }
```

kasobol-msft replied:
Ahh. I didn't realize I'm looking at that generated file. my bad.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/tests/CryptoraphyTestsExtensionMethods.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Azure.Storage.Blobs.Cryptography.Tests
+{
+    internal static class CryptoraphyTestsExtensionMethods
```
Addressed in next push.

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/WindowStream.cs:
```diff
@@ -0,0 +1,81 @@
+            throw new NotImplementedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
```
Addressed in next push.

jaschrep-msft committed:
Queues handles partial decryption errors.

Queues now has a listener for when only some messages cannot be
decrypted. these messages are filtered out of the response and sent to
the listened instead. If no listener is provided, the whole fetch
throws.
Some PR comments addressed.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/AlwaysFailsKeyEncryptionKeyResolver.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace Azure.Storage.Blobs.Tests
+{
+    internal class AlwaysFailsKeyEncryptionKeyResolver : IKeyEncryptionKeyResolver
```
nit: it doesn't always fail.

Btw. I'd use Moq for such simple case rather than creating new type. I believe we have it as dependency.

jaschrep-msft replied:
addressed in upcoming push

kasobol-msft commented on sdk/storage/Azure.Storage.Queues/tests/MockMissingClientSideEncryptionKeyListener.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Azure.Storage.Queues.Tests
+{
+    internal class MockMissingClientSideEncryptionKeyListener : IMissingClientSideEncryptionKeyListener
```
I'd use Moq for that as well.

jaschrep-msft replied:
addressed in upcoming push

kasobol-msft reviewed (commented):
Looks good for shipping preview from it. We should address rest of the feedback here or before GA.

jaschrep-msft committed:
Removed nonexistent project from sln

jaschrep-msft committed:
Change in handling key resolution

Clientside decryption will now either succeed or throw. Queues can
redirect their throw to a listener.

jaschrep-msft committed:
Many PR comments addressed; key-substitution added

jaschrep-msft committed:
Refactors; Mocking some tests

jaschrep-msft committed:
Refactors and bug fixes

jaschrep-msft committed:
Export-API

gapra-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -1040,13 +1047,19 @@ public static partial class SpecializedBlobExtensions
+        {
+            if (client.ClientSideEncryption != default)
+            {
+                throw Errors.ClientSideEncryption.TypeNotSupported(typeof(BlockBlobClient));
```
Should this be typeof(AppendBlobClient)? I think this was also missed in PageBlobClient

jaschrep-msft committed:
Constant/error/assertion refactors

gapra-msft reviewed (approved):
Looks good to me

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:
```diff
@@ -1020,6 +1028,26 @@ public BlobClient(Uri blobUri, TokenCredential credential, BlobClientOptions opt
             bool async = true,
             CancellationToken cancellationToken = default)
         {
+            // TODO uncomment when upload from file gets its own implementation
```
to be addressed in this PR?

jaschrep-msft replied:
No. This is to do with future work to give upload from file its own implementation separate from upload from stream. However, such an implementation will be incompatible with this feature, and so it will need to revert to upload from stream. I can clarify this.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -686,6 +709,16 @@ private void SetNameFieldsIfNull()
+                    // we already return a nonseekable stream; returning a crypto stream is fine
+                    if (UsingClientSideEncryption)
+                    {
+                        stream = await new BlobClientSideDecryptor(ClientSideEncryption)
+                            .ClientSideDecryptInternal(stream, response.Value.Metadata, requestedRange, response.Value.ContentRange, async, cancellationToken).ConfigureAwait(false);
```
Two things to consider:
- It seems that `BlobClientSideDecryptor` is stateless in this context. Could be a member (field) in this class instead of storing encryption options.
- It's not far away from here to create `IReadStreamTransformer` and have `BlobClientSideDecryptor : IReadStreamTransformer`

not blocking here, but would be nice to have.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:
```diff
@@ -946,6 +947,13 @@ public BlobClient(Uri blobUri, TokenCredential credential, BlobClientOptions opt
+            if (UsingClientSideEncryption)
+            {
+                // content is now unseekable, so PartitionedUploader will be forced to do a buffered multipart upload
+                (content, metadata) = await new BlobClientSideEncryptor(ClientSideEncryption)
+                    .ClientSideEncryptInternal(content, metadata, async, cancellationToken).ConfigureAwait(false);
```
Similar feedback here as for Decryptor.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClientSideEncryptor.cs:
```diff
@@ -0,0 +1,78 @@
+                async,
+                cancellationToken).ConfigureAwait(false);
+
+            //Stream seekableCiphertext = new RollingBufferStream(
+            //        nonSeekableCiphertext,
+            //        EncryptionConstants.DefaultRollingBufferSize,
+            //        GetExpectedCryptoStreamLength(originalLength));
```
Please cleanup commented out code here and there. Either we should work items for these improvements or implement them right away.

jaschrep-msft replied:
This was old code that got moved around in the refactors. Is obsolete and will be removed.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlockBlobClient.cs:
```diff
@@ -153,6 +153,7 @@ public BlockBlobClient(string connectionString, string containerName, string blo
         public BlockBlobClient(string connectionString, string blobContainerName, string blobName, BlobClientOptions options)
             : base(connectionString, blobContainerName, blobName, options)
         {
+            AssertNoClientSideEncryption(options);
```
similar feedback here. let's try to consolidate this logic.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/PageBlobClient.cs:
```diff
@@ -98,6 +99,7 @@ public PageBlobClient(string connectionString, string blobContainerName, string
         public PageBlobClient(string connectionString, string blobContainerName, string blobName, BlobClientOptions options)
             : base(connectionString, blobContainerName, blobName, options)
         {
+            AssertNoClientSideEncryption(options);
```
and here.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/tests/ClientSideEncryptionTests.cs:
```diff
@@ -0,0 +1,673 @@
+        private Mock<IKeyEncryptionKeyResolver> GetAlwaysFailsKeyResolver(bool throws)
+        {
+            var mock = new Mock<IKeyEncryptionKeyResolver>(MockBehavior.Strict);
+            if (IsAsync)
```
I'd consider removing these kind of branches and just mock both sync and async in one shot. that'd be less screens of code.
Unless we want to verify that right method was called.

jaschrep-msft replied:
We do want to verify the right method was called. All our tests only mock the correct method and set strict mock behavior. This way, we can verify a change didn't mess up our sync/async story, which can be easy to miss.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/ClientSideDecryptor.cs:
```diff
@@ -0,0 +1,253 @@
+        /// Exceptions thrown based on implementations of <see cref="IKeyEncryptionKey"/> and
+        /// <see cref="IKeyEncryptionKeyResolver"/>.
+        /// </exception>
+        public static async Task<Stream> DecryptInternal(
+            Stream ciphertext,
+            EncryptionData encryptionData,
+            bool ivInStream,
+            IKeyEncryptionKeyResolver keyResolver,
+            IKeyEncryptionKey potentialCachedKeyWrapper,
+            bool noPadding,
+            bool async,
+            CancellationToken cancellationToken)
```
I'd rather model this class in same way `BlobClientSideDecryptor` is. `IKeyEncryptionKey` and `IKeyEncryptionKeyResolver` are invariant for `DecryptInternal` so they're more of class dependencies rather than input. `BlobClientSideDecryptor` could take `ClientSideDecryptor` as a dependency.

There are also benefits of killing static code, see https://objcsharp.wordpress.com/2013/07/08/why-static-code-is-bad/ for more detailed explanation. It's a cosmetic change now, can be a headache in long run.

jaschrep-msft replied:
Agreed. I believe the package-specific instance decryptors are better. And this would help the possible future where more modes of encryption are added.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/ClientSideEncryptor.cs:
```diff
@@ -0,0 +1,124 @@
+        /// <param name="async">Whether to wrap the CEK asynchronously.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>The wrapped stream to read from and the encryption metadata for the wrapped stream.</returns>
+        public static async Task<(Stream ciphertext, EncryptionData encryptionData)> EncryptInternal(
```
similar feedback here.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionDataSerializer.cs:
```diff
@@ -0,0 +1,196 @@
+        /// </summary>
+        /// <param name="data">Data to serialize.</param>
+        /// <returns>The JSON UTF8 bytes.</returns>
+        public static ReadOnlyMemory<byte> SerializeEncryptionData(EncryptionData data)
```
does this have to be public ?

jaschrep-msft replied:
It's on an internal class.

kasobol-msft replied:
I mean we call it from method above. Do we call it from somewhere else? if not then we could hide it more.

kasobol-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClientExtensions.cs:
```diff
@@ -0,0 +1,43 @@
+        /// <param name="client">Client to base off of.</param>
+        /// <param name="listener">Listener for when decryption of a single message fails.</param>
+        /// <returns>New instance with provided options and same internals otherwise.</returns>
+        public static QueueClient WithClientSideEncryptionFailureListener(this QueueClient client, IClientSideDecryptionFailureListener listener)
```
If it's possible to have extension method taking optional parameters then I'd consider consolidating these extension into `WithClientSideEncryption(this client, options, listener = default)`.

jaschrep-msft replied:
I originally had this and decided against it. There is no easy way to distinguish between leaving one of the objects as is versus clearing it. I'd rather have the customer set each separately than force them to maintain a reference to one when they only want to change the other. Since the whole point of this method is to provide a lightweight mechanism to swap the key, I am not worried about someone who actually wants to change both having to call into 2 separate lightweight methods.

kasobol-msft replied:
gotcha. so we want to prevent client.WithClientEcryption(options, litenere).WithClientEncryption(options) from zeroing listener. Makes sense.

kasobol-msft commented on sdk/storage/Azure.Storage.Queues/tests/EncryptedMessageSerializerTests.cs:
```diff
@@ -0,0 +1,181 @@
+        [TestCase("")]
+        [TestCase("\"aa\"")] // real world example
+        [TestCase("this is not even valid json")]
+        [TestCase("ᛁᚳ᛫ᛗᚨᚷ᛫ᚷᛚᚨᛋ᛫ᛖᚩᛏᚪᚾ᛫ᚩᚾᛞ᛫ᚻᛁᛏ᛫ᚾᛖ᛫ᚻᛖᚪᚱᛗᛁᚪᚧ᛫ᛗᛖ")]
```
what alphabet is that ?

jaschrep-msft replied:
Anglo-Saxon runes.

kasobol-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionDataSerializer.cs:
```diff
@@ -0,0 +1,196 @@
+
+namespace Azure.Storage.Cryptography.Models
+{
+    internal static class EncryptionDataSerializer
```
Seeing so much static code here is little bit itchy (see comments on the Encryptor/Decryptor in common). However in this case we don't seem to have class dependencies and we don't have DI container and it's how we do serialization/deserialization everywhere else, so I'm going to pass on this here.

kasobol-msft reviewed (commented):
It's much easier to navigate this code now. Thank you!

Please take a look at the comment I left on QueueClient crypto extension and see if you want to do it before providing another drop.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -98,6 +98,7 @@ public AppendBlobClient(string connectionString, string blobContainerName, strin
         public AppendBlobClient(string connectionString, string blobContainerName, string blobName, BlobClientOptions options)
             : base(connectionString, blobContainerName, blobName, options)
         {
+            AssertNoClientSideEncryption(options);
```
I'd be worried doing that, as we then have to repeat any transformations we make in BlobBaseClient going from one overload to another. Perhaps we can have a single check method in BlobBaseClient that's a no-op and have block, page, and append clients override the check to actually make the assertion?

kasobol-msft replied:
ah, that's the catch. let's leave it then.

jaschrep-msft committed:
Refactors

jaschrep-msft commented:
Unsure how Microsoft.Extensions.Azure got pulled in here. Will revert.

jaschrep-msft committed:
Revert deletion of an api.cs file

tg-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -989,6 +989,10 @@ public partial class BlobBaseClient
         public virtual Azure.Storage.Blobs.Specialized.BlobBaseClient WithSnapshot(string snapshot) { throw null; }
         protected virtual Azure.Storage.Blobs.Specialized.BlobBaseClient WithSnapshotCore(string snapshot) { throw null; }
     }
+    public static partial class BlobClientExtensions
```
This should go in [SpecializedBlobExtensions](https://apiview.dev/Assemblies/Review/a17b6175b93847db83a99c7fcfeca9ab#Azure.Storage.Blobs.Specialized.SpecializedBlobExtensions) so we don't add a new type.

jaschrep-msft replied:
I'll mirror this in Queues as well. There needed to be a new type no matter what, but now it will be in line with this Blobs change.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:
```diff
@@ -1037,6 +1041,11 @@ public partial class BlockBlobClient : Azure.Storage.Blobs.Specialized.BlobBaseC
         public new Azure.Storage.Blobs.Specialized.BlockBlobClient WithSnapshot(string snapshot) { throw null; }
         protected sealed override Azure.Storage.Blobs.Specialized.BlobBaseClient WithSnapshotCore(string snapshot) { throw null; }
     }
+    public partial class ExtendedBlobClientOptions : Azure.Storage.Blobs.BlobClientOptions
```
`Extended` is probably not the best prefix because a lot of folks have a visceral reaction to it.  Could we call this `SpecializedBlobClientOptions` to match the `Specialized` namespace and terminology we've been building?

tg-msft reviewed (commented):
Commenting on the APIs first as they're the most important.  I'll run through the rest as well.

weshaggard commented on sdk/storage/Azure.Storage.Queues/tests/Azure.Storage.Queues.Tests.csproj:
```diff
@@ -13,6 +13,9 @@
     <PackageReference Include="Azure.Identity" />
+    <PackageReference Include="Azure.Security.KeyVault.Keys" />
+    <PackageReference Include="Microsoft.Azure.KeyVault.Core" />
+    <PackageReference Include="Microsoft.Azure.Storage.Queue" />
```
Do we really want to reference the track1 version of these libraries?

jaschrep-msft replied:
Yes. There are compatibility tests to ensure each library can read the data encoded by the other. This is an upgrade requirement for anyone who used client-side encryption in track 1.

kasobol-msft replied:
better safe than sorry. We can think about spinning extra test package(s) where we test track1 track2 interop aspects in long run or take it as follow up from here.

weshaggard replied:
Yes I do think it is worth considering multiple test projects to help ensure we don't have any weird cross-pollination between the track 1 and track 2 library.

jaschrep-msft committed:
Minor Crypto API Adjustments

jaschrep-msft committed:
Export API

jaschrep-msft commented:
/azp run net storage - ci

jaschrep-msft commented:
/azp run net - storage - ci

tg-msft reviewed (commented):
I'm adding a bunch more comments, but only things affecting the API should be worried about for Preview.  The rest can be addressed in future changes if you think they're worth doing.

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/Models/EncryptedMessage.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    internal class EncryptedMessage
+    {
+        public string EncryptedMessageContents { get; set; }
```
Should this be `EncryptedMessageText` to map to `QueueMessage.MessageText`?

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/Models/EncryptedMessageSerializer.cs:
```diff
@@ -0,0 +1,95 @@
+
+namespace Azure.Storage.Queues.Specialized.Models
+{
+    internal static class EncryptedMessageSerializer
```
Nit - Any reason not to put these in the `EncryptedMessage` class?

jaschrep-msft replied:
Do you mean placing the serializer as a nested class to what it is serializing?

tg-msft replied:
No, just merging the two classes together.

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClientSideDecryptor.cs:
```diff
@@ -0,0 +1,102 @@
+                }
+                catch (Exception e) when (_listener != default)
+                {
+                    if (async)
```
This pattern doesn't feel right.  What if I implemented a sync listener becuase I only Console.WriteLine but use QueueClient.ReceiveAsync to make an async method call?

jaschrep-msft replied:
Will come back to this based on discussion of what form the listener should even take.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -370,7 +387,7 @@ internal BlobBaseClient(Uri blobUri, HttpPipelinePolicy authentication, BlobClie
         {
             var builder = new BlobUriBuilder(Uri) { Snapshot = snapshot };
-            return new BlobBaseClient(builder.ToUri(), Pipeline, Version, ClientDiagnostics, CustomerProvidedKey, EncryptionScope);
+            return new BlobBaseClient(builder.ToUri(), Pipeline, Version, ClientDiagnostics, CustomerProvidedKey, ClientSideEncryption, EncryptionScope);
```
That is in the pipeline of work this coming semester.

kasobol-msft replied:
I like that idea

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClientSideDecryptor.cs:
```diff
@@ -0,0 +1,102 @@
+                    }
+                }
+            }
+            return filteredMessages.ToArray();
```
So you have to implement a listener if you don't want to wait for the visiblity timeout on messages you can't read?  This design feels tricky to use correctly.

jaschrep-msft replied:
The alternatives as far as we've explored are to
1. Return the serialized encrypted message schema and let the user write the code to discover they failed to decrypt.
2. Silently filter out messages we couldn't decrypt.

I'm not happy with either of those. The first is undue burden on the customer, whereas through the listener we do the determination for them and they handle it how they wish. The second feels dangerous in that queue messages can just time out with no one noticing.

I suppose an option 3 could exist of overriding the failed messages to throw upon accessing the message text and just always return all the messages, but I'm not sure whether that's useful.

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueServiceClient.cs:
```diff
@@ -59,6 +61,18 @@ public class QueueServiceClient
+        /// </summary>
+        internal virtual ClientSideEncryptionOptions ClientSideEncryption => _clientSideEncryption;
+
+        private readonly IClientSideDecryptionFailureListener _missingClientSideEncryptionKeyListener;
```
Nit - this name doesn't match the type like the other uses.

tg-msft reviewed (commented):
Adding a few more comments

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionAgent.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// The protocol version used for encryption.
+        /// </summary>
+        public ClientSideEncryptionVersion Protocol { get; set; }
```
Nit - I'd suggest naming this `EncryptionVersion` instead of `Protocol` to match the name in Options.

kasobol-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClientSideDecryptor.cs:
```diff
@@ -0,0 +1,183 @@
+
+namespace Azure.Storage.Blobs
+{
+    internal class BlobClientSideDecryptor
```
@tg-msft current naming might indicate that both implement same interface (looks like decorators/wrappers) little bit.

However, I'd rather not take it back to the static code. See previous discussion https://github.com/Azure/azure-sdk-for-net/pull/12183#discussion_r434152869

I was suggesting earlier to make BlobClientSideDecryptor a readonly field and populate if encryption is enabled (it's stateless within the BlobClient lifetime).

tg-msft replied:
I'm fine with whatever you all want here.  I just struggled to understand what was happening at first with these names.

jaschrep-msft commented on sdk/storage/Azure.Storage.Queues/api/Azure.Storage.Queues.netstandard2.0.cs:
```diff
@@ -315,6 +315,27 @@ public partial class UpdateReceipt
+        void OnFailure(Azure.Storage.Queues.Models.QueueMessage message, System.Exception exception);
+        System.Threading.Tasks.Task OnFailureAsync(Azure.Storage.Queues.Models.PeekedMessage message, System.Exception exception);
+        System.Threading.Tasks.Task OnFailureAsync(Azure.Storage.Queues.Models.QueueMessage message, System.Exception exception);
+    }
```
If we believe we don't need async methods here then so be it. But we have two distinct types of QueueMessage and PeekedMessage that don't have a common base. Is accepting more than one action when the user probably wants to always supply both really the best option? I'm not sure of a better way to supply this functionality.

tg-msft replied:
@KrzysztofCwalina and I talked about this a bit and came up with the following to match what we did for QuickQuery a little better here:

```C#
    public class QueueClientSideEncryptionOptions : ClientSideEncryptionOptions
    {
        public event EventHandler<DecryptionFailureEventArgs> DecryptionFailed;
    }

    public class ClientSideDecryptionFailureEventArgs
    {
        public Exception Exception { get; }
        public ClientSideDecryptionFailureEventArgs(Exception exception);
    }
```

and you could use it like:

```C#
var options = new QueueClientSideEncryptionOptions();
options.KeyResolver = ...;
options.DecryptionFailed +=
    (object sender, ClientSideDecryptionFailureEventArgs args) =>
    {
        if (sender is PeekedMessage peek)
        {
             Console.WriteLine($"Failed to peek at {peek.MessageId}");
        }
        else if (sender is QueueMessage message)
        {
             Console.WriteLine($"Failed to receive {message.MessageId}");
        }
    };
```

It's only sync for now, but so is QuickQuery's error handler and I'd wait to hear feedback from folks if they need an async event.

Do you this this would work?

jaschrep-msft replied:
Is the suggestion that `SpecializedQueueClientOptions` now have a property for `QueueClientSideEncryptionOptions` instead of the normal `ClientSideEncryptionOptions`, or that we still accept the base class and work it through internals like we do with client options in the first place?

tg-msft replied:
I'd say leave it as the base `ClientSideEncryptionOptions` in `SpecializedQueueClientOptions ` so people could pass the same value for Blobs and Queues.  You'd only need to learn about `QueueClientSideEncryptionOptions` if you were trying to handle failures which should be an advanced scenario.

On a related note - maybe we should change the semantics to throw an exception when you can't decrypt if there's no failure handler?  Then new users who aren't trying to use multiple keys/etc. will still have an easy notification that something went wrong but advanced scenarios are still possible for special partners?

tg-msft replied:
And you could just do an `if (options is QueueClientSideEncryptionOptions advanced) { _failureHandler = advanced.DecryptionFailed; }` instead of a bunch of internal state sharing since `ClientSideEncryptionOptions` is in Common.

jaschrep-msft commented on sdk/storage/Azure.Storage.Queues/api/Azure.Storage.Queues.netstandard2.0.cs:
```diff
@@ -315,6 +315,27 @@ public partial class UpdateReceipt
+    }
+    public static partial class SpecializedQueueExtensions
+    {
+        public static Azure.Storage.Queues.QueueClient WithClientSideEncryptionFailureListener(this Azure.Storage.Queues.QueueClient client, Azure.Storage.Queues.Specialized.IClientSideDecryptionFailureListener listener) { throw null; }
```
It seemed to go hand in hand with replacing the key. Private drops have been given with this functionality, so we'll have to check with our partners on whether they actually use this if we think it should be removed.

tg-msft replied:
If we move to the `EventHandler<T>` approach, hopefully the single `With()` method will suffice.

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,89 @@
+        /// <summary>
+        /// The blob encryption mode.
+        /// </summary>
+        public string EncryptionMode { get; set; }
```
Nit - should things like this be settable since they're effectively constants in v1?

jaschrep-msft replied:
It's part of the JSON schema. We need to be able to parse the value and verify it.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/AppendBlobClient.cs:
```diff
@@ -192,9 +196,24 @@ public AppendBlobClient(Uri blobUri, TokenCredential credential, BlobClientOptio
+                  clientSideEncryption: default,
+                  encryptionScope)
         {
         }
```
There aren't any encryption options being passed in here. It's definitional that they don't exist. Are you suggesting we add the encryption options to the constructor explicitly even though we know we will never use them?

I know outside this comment you mentioned a complete refactoring of how our constructors are handled internally, which I mentioned is already an upcoming work item. Are you okay with this being addressed during that work?

tg-msft replied:
I'm okay with anything but the API changes happening as future work.

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Shared/ClientsideEncryption/ClientSideEncryptionOptionsExtensions.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    internal static class ClientSideEncryptionOptionsExtensions
+    {
+        public static ClientSideEncryptionOptions Clone(this ClientSideEncryptionOptions options)
```
Nit - maybe add a mainteance note in ClientSideEncryptionOptions that this msut be updated when changes are introduced there?

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -325,20 +339,23 @@ internal BlobBaseClient(Uri blobUri, HttpPipelinePolicy authentication, BlobClie
             ClientDiagnostics clientDiagnostics,
             CustomerProvidedKey? customerProvidedKey,
+            ClientSideEncryptionOptions clientSideEncryption,
             string encryptionScope)
```
Are you suggesting a new internal class that just combines all encryption-related options for passing them between clients? If we need a new internal constructor story anyway, I'm not sure what this buys us.

tg-msft replied:
No, I'm just talking about parameter ordering.  It's a super nit comment.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClientSideDecryptor.cs:
```diff
@@ -0,0 +1,183 @@
+            _decryptor = decryptor;
+        }
+
+        public async Task<Stream> ClientSideDecryptInternal(
```
Will call it `DecryptInternal` to match patterns of `Internal` suffix on async branching functions.

tg-msft reviewed (approved):
Adding a few last comments.  I'm approving and it's fine if you want to merge this now and continue addressing the last API feedback in a follow up PR.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClientSideDecryptor.cs:
```diff
@@ -0,0 +1,183 @@
+            return await TrimStreamInternal(plaintext, originalRange, contentRange, ivInStream, async, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<Stream> TrimStreamInternal(Stream stream, HttpRange originalRange, ContentRange? receivedRange, bool pulledOutIv, bool async, CancellationToken cancellationToken)
```
The above method does a computation and passes true or false as the result of that computation. Will do the multi-line formatting.

jaschrep-msft committed:
Queue listener changed to events.

Other minor PR feedback.

jaschrep-msft committed:
Export api

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClientSideEncryptionOptions.cs:
```diff
@@ -0,0 +1,104 @@
+
+        internal void OnDecryptionFailed(object message, Exception e)
+        {
+            DecryptionFailed?.Invoke(this, new ClientSideDecryptionFailureEventArgs(message, e));
```
We should pass `message` as the source instead of `this` (i.e., it was the message that failed to decrypt).  This also kind of hides the wart of dealing with QueueMessage vs PeekedMessage behind the event handler pattern that forces `object sender`.

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClientSideEncryptionOptions.cs:
```diff
@@ -0,0 +1,104 @@
+        /// Message the failure occured with. Can be an instance of either
+        /// <see cref="Queues.Models.QueueMessage"/> or <see cref="Queues.Models.PeekedMessage"/>.
+        /// </summary>
+        public object Message { get; }
```
This should be removed as well if we're passing it as the event's sender.

tg-msft commented on sdk/storage/Azure.Storage.Queues/src/QueueClientSideEncryptionOptions.cs:
```diff
@@ -0,0 +1,104 @@
+        /// Compiler restriction: can only copy an event in an instance method on the class containing the event.
+        /// This class exists to allow us to copy the event.
+        /// </remarks>
+        private QueueClientSideEncryptionOptions CloneAsQueueClientSideEncryptionOptions()
+        {
+            // clone base class but as instance of this class
+            var newOptions = new QueueClientSideEncryptionOptions(EncryptionVersion);
+            ClientSideEncryptionOptionsExtensions.CopyOptions(this, newOptions);
+
+            // clone data specific to this subclass
+            newOptions.DecryptionFailed = DecryptionFailed;
+
+            return newOptions;
+        }
+
+        /// <summary>
+        /// Clones the given <see cref="ClientSideEncryptionOptions"/> as an instance of
+        /// <see cref="QueueClientSideEncryptionOptions"/>. If the given instance is also a
+        /// <see cref="QueueClientSideEncryptionOptions"/>, this clones it's specialty data as well.
+        /// </summary>
+        /// <returns></returns>
+        internal static QueueClientSideEncryptionOptions CloneFrom(ClientSideEncryptionOptions options)
+        {
+            if (options == default)
+            {
+                return default;
+            }
+            else if (options is QueueClientSideEncryptionOptions)
+            {
+                return ((QueueClientSideEncryptionOptions)options).CloneAsQueueClientSideEncryptionOptions();
+            }
+            else
+            {
+                // clone base class but as instance of this class
+                var newOptions = new QueueClientSideEncryptionOptions(options.EncryptionVersion);
+                ClientSideEncryptionOptionsExtensions.CopyOptions(options, newOptions);
+
+                return newOptions;
+            }
+        }
```
Nit - could we simplify both methods to just?

```C#
internal static QueueClientSideEncryptionOptions CloneFrom(ClientSideEncryptionOptions options)
{
    if (options == default) { return default; }
    var newOptions = new QueueClientSideEncryptionOptions(options.EncryptionVersion);
    ClientSideEncryptionOptionsExtensions.CopyOptions(options, newOptions);
    if (options is QueueClientSideEncryptionOptions queueOptions)
    {
        newOptions.DecryptionFailed = queueOptions.DecryptionFailed;
	}
    return newOptions;
}
```

jaschrep-msft replied:
Huh. Somehow that gets past the compiler restrictions on when an event can be assigned. I did something nearly identical before and it said no. Will fix.

tg-msft reviewed (approved):
There's one small thing we should change about the event source, but it otherwise looks great to me.  :shipit:

jaschrep-msft committed:
Testing for update message encryption

jaschrep-msft committed:
Minor event API adjustments

jaschrep-msft committed:
export api

jaschrep-msft commented:
/azp run net - storage - ci

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into feature/storage/integrated-clientside-encryption

tg-msft commented:
/check-enforcer reset

tg-msft commented:
/check-enforcer evaluate
